### ISSUE TITLE
Devops does not start container-app-job for self-hosted agent

### ISSUE THREAD

**Issue by MrTomato** [microsoft/azure-container-apps#1064]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is a: (mark with an x)
- [x] bug report -> please search issues before submitting
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

### Issue description
Hi team,

I set up the self-hosted agent to run inside Container-App-Job being triggered for builds.
The tutorial I followed was https://learn.microsoft.com/en-gb/azure/container-apps/tutorial-ci-cd-runners-jobs?tabs=bash&pivots=container-apps-jobs-self-hosted-ci-cd-azure-pipelines.

I can start a job manually and it will register with DevOps’ agent-pool.
Then I can also run builds on that agent-pool.
Therefore I assume all of the connecting works and credentials are fine.

However, if I just start a pipeline run from inside DevOps without manually starting a container-app-job before, the build fails with

##[error]No agent found in pool container-apps-tmp which satisfies the specified demands: Agent.Version -gtVersion 2.163.1

I tried to see how DevOps would notify the Container-App-Job of a pending build, but I couldn’t find any details and therefore couldn’t debug it myself.

Please advise how I could find out where the events should be published and how I might find why the job is not being triggered.

Thanks,
Markus

### Steps to reproduce

Execute the following steps in bash (I censored the confidential parts, details can be provided to MS support):

``` shell
stage="dev"
az account set -s "*****-$stage"

CONTAINER_APP_ENVIRONMENT="container-apps-tmp"
CONTAINER_APP_JOB="container-apps-tmp"
DEVOPS_AGENT_POOL="container-apps-tmp"

DEVOPS_AGENT_POOL_PAT="The PAT with Agent Pools (Read & manage) permissions"
ACR_PWD="The pwd for a user that is allowed to pull the self-hosted-agent image from ACR"

LOCATION="westeurope"
RESOURCE_GROUP="A fresh resource group name"

CONTAINER_REGISTRY="******"
CONTAINER_IMAGE="******"
ACR_USER="********"

DEVOPS_URL="https://dev.azure.com/*******"

az group create \
    --name "$RESOURCE_GROUP" \
    --location "$LOCATION"

az containerapp env create \
    --name "$CONTAINER_APP_ENVIRONMENT" \
    --resource-group "$RESOURCE_GROUP" \
    --location "$LOCATION"

az containerapp job create -n "$CONTAINER_APP_JOB" -g "$RESOURCE_GROUP" --environment "$CONTAINER_APP_ENVIRONMENT" \
    --trigger-type Event \
    --replica-timeout 1800 \
    --replica-retry-limit 0 \
    --replica-completion-count 1 \
    --parallelism 1 \
    --image "$CONTAINER_REGISTRY.azurecr.io/$CONTAINER_IMAGE" \
    --min-executions 0 \
    --max-executions 10 \
    --polling-interval 5 \
    --scale-rule-name "azure-pipelines" \
    --scale-rule-type "azure-pipelines" \
    --scale-rule-metadata "poolName=$DEVOPS_AGENT_POOL" "targetPipelinesQueueLength=1" \
    --scale-rule-auth "personalAccessToken=personal-access-token" "organizationURL=organization-url" \
    --cpu "2.0" \
    --memory "4Gi" \
    --secrets "personal-access-token=$DEVOPS_AGENT_POOL_PAT" "organization-url=$DEVOPS_URL" \
    --env-vars "AZP_TOKEN=secretref:personal-access-token" "AZP_URL=secretref:organization-url" "AZP_POOL=$DEVOPS_AGENT_POOL" \
    --registry-server "$CONTAINER_REGISTRY.azurecr.io" \
    --registry-username "$ACR_USER" \
    --registry-password "$ACR_PWD"

```

Then go into DevOps and create a new self-hosted agent-pool where with the name "container-apps-tmp".

Now create a new "hello-world" pipeline using that agent pool:

```
pool:
  name: container-apps-tmp

steps:
  - script: echo Hello, world!
     displayName: 'Run a one-line script'
```

and press run to execute the build pipeline.

**Expected behavior** [What you expected to happen.]

A new build is triggered. 
A new container job for the agent is spun up. 
The build starts executing.

**Actual behavior** [What actually happened.]

A new build is triggered and fails immediately with

```
##[error]No agent found in pool container-apps-tmp which satisfies the specified demands: Agent.Version -gtVersion 2.163.1
Pool: [container-apps-tmp](https://dev.azure.com/*********/f713cc9c-02e4-4034-a05b-0bd9934f14b0/_settings/agentqueues?poolId=&queueId=86)
Started: Just now
Duration: 10s
```

### Additional context

I already asked the Devops folks for support, but they redirect me to you. 

The command I used to manually start the agent (and then have it register into Devops and run my builds) was:

``` shell
# Manually start the image

az containerapp job start -n "$CONTAINER_APP_JOB" -g "$RESOURCE_GROUP"
az containerapp job execution list \
    --name "$CONTAINER_APP_JOB" \
    --resource-group "$RESOURCE_GROUP" \
    --output table \
    --query '[].{Status: properties.status, Name: name, StartTime: properties.startTime}'
```

**Comment by vinisoto** [comment]
the pipeline might be adding this demand `Agent.Version -gtVersion 2.163.1` behind the scenes. Can you specify Agent.Version = 2.163.***2*** as a capability of your pool?

**Comment by MrTomato** [comment]
@vinisoto the issue is that the pool does not contain any agents that have any capabilities associated, since no agent is started.
If I start the container job manually and an agent registers, the build is scheduled i.e. the demands are matched.

Therefore I would assume this is just the first check that runs to see if there are any ready agents. This is of course not fulfilled if there are no running agents.

**Comment by gabesmsft** [comment]
@MrTomato - is the [placeholder agent](https://learn.microsoft.com/azure/container-apps/tutorial-ci-cd-runners-jobs?tabs=bash&pivots=container-apps-jobs-self-hosted-ci-cd-azure-pipelines#create-a-placeholder-self-hosted-agent) present in the Agent pool? The error will occur if the placeholder agent isn't present. Although you can delete the Container App Job that creates the placeholder agent, the placeholder agent itself must remain present in the agent pool.

**Comment by gabesmsft** [comment]
@qriticual - for me, it works with the offline placeholder agent. To verify that something else in the system wasn't causing the pipeline to run without errors: I deleted my offline placeholder agent and triggered a pipeline run (which failed as expected due to no agent), re-ran the placeholder agent creation via the Job (which created an offline placeholder agent), and then re-ran the pipeline (which succeeded as expected).

I am not sure why it isn't working for you when you have the offline agent. On your offline pipeline agent, is the Agent.OS capability present and set to Linux, and is the Agent.Version capabilty present and set to a version that is higher than 2.163.1 (or more generally, higher than whichever version is mentioned in the ADO error)? If so, it should succeed for demands in the specified agent pool.

**Comment by qriticual** [comment]
@gabesmsft - I don't know why it doesn't work for me.. It works fine when having a container app that runs all the time, however it's a costlier option therefore I want the event driven app job. The pipeline is just stuck waiting for an agent to be available but the event job never starts. I have deployed it using terraform. After that I've created a place holder agent with a manual job. The Agent.OS for the placeholder agent is Linux and Agent.Version is 3.234.0. This is my terraform code for reference: `resource "azapi_resource" "adojob" {
  type      = "Microsoft.App/jobs@2023-05-01"
  name      = local.aca_name
  location  = var.location
  parent_id = var.parent_id
  identity {
    type = "UserAssigned"
    identity_ids = [
      azurerm_user_assigned_identity.aca.id
    ]
  }
  body = jsonencode({
    properties = {
      environmentId = azurerm_container_app_environment.ado.id
      configuration = {
        registries = [
          {
            identity = azurerm_user_assigned_identity.aca.id
            server   = var.container_registry
          }
        ]
        triggerType = "Event"
        replicaTimeout    = 30
        replicaRetryLimit = 1
        eventTriggerConfig = {
          replicaCompletionCount = 1
          parallelism            = 1
          scale = {
            minExecutions   = 0
            maxExecutions   = 10
            pollingInterval = 5
            rules = [
              {
                name = "azure-pipelines",
                type = "azure-pipelines",
                metadata = {
                  "poolID"                      = var.container.ado_pool_id
                  "poolName"                    = var.container.ado_pool_name
                  "targtetPipelinesQueueLength" = "1"
                }
                auth = [
                  {
                    secretRef         = "personal-access-token"
                    triggerParameter  = "personalAccessToken"
                  },
                  {
                    secretRef         = "organization-url"
                    triggerParameter  = "organizationURL" 
                  }
                ]
              }
            ]

          }
        }
        secrets = [
          {
            name  = "personal-access-token"
            value = var.container.ado_pat
          },
          {
            name  = "organization-url"
            value = var.container.ado_url
          }
        ]
      }
      template = {
        containers = [
          {
            name  = var.container.name
            image = var.container.image
            env = [
              {
                name      = "AZP_TOKEN"
                secretRef = "personal-access-token"
              },
              {
                name      = "AZP_URL"
                secretRef = "organization-url"
              },
              {
                name      = "AZP_POOL"
                value     = "${var.container.ado_pool_name}"
              },
              {
                name      = "AZP_POOLID"
                value     = "${var.container.ado_pool_id}"
              }
            ]
            resources = {
              cpu    = var.container.cpu
              memory = var.container.memory
            }
          }
        ]
      }
    }
  })
  depends_on = [azurerm_role_assignment.acr]
}`
And here is the capabilities of the placeholder agent: 
![agent_capabilities](https://github.com/microsoft/azure-container-apps/assets/62610168/c8a2b5ba-bf3f-4ad3-b892-e60f5fe5c700)

**Comment by MrTomato** [comment]
After further analysis and a coworker following the tutorial to the letter: We are using a ACR that is in a different subscription.
I'm finding it hard to believe that that's the issue, but it seems us like that is the issue.

The subscriptions belong to the same tenant and the specified registry user has access to it.

Can you reproduce this @gabesmsft ?

**Comment by gabesmsft** [comment]
I apologize about the delay in responding. For me, the placeholder agent was running without issue recently, but I am suddenly encountering an error of "Cannot connect to server, because config files are missing".
I'm checking with some folks about what could be causing this error. I don't know if it is the same issue(s) that each of you are facing though. 

Could you [check the execution history](https://learn.microsoft.com/azure/container-apps/jobs?tabs=azure-portal) of the Job, to verify whether it failed? If it failed, click the Console link next to the failure to check the logs for the execution and see if there are any error messages?
@MrTomato - could you check these logs for the placeholder ACA job that is having the issue? I will note that the agent is supposed to show as offline in ADO, but it is also supposed to remain there (and not disappear), and it seems that the timeout that you are mentioning is worth checking the logs for.

@qriticual - could you check these logs for the agent ACA job that is supposed to run? If the ADO pipeline is waiting for an agent to become available, it could be that the job is crashing, for which the logs would hopefully give further insight into. The objective is definitely to have an on-demand job rather than a constantly running ACA for this scenario.


@MrTomato I'm thinking the different subscription isn't an issue, since running the Job causes some ADO agent-specific activity to happen (i.e. it put the placeholder agent in ADO and (at least in my case) generated some ADO agent-specific logs, which wouldn't happen if the Job couldn't pull the image from ACR). But I'm encountering my error even within the same subscription, so I don't have a definitive test for the different sub case yet.

**Comment by MrTomato** [comment]
@gabemsft we finally found the issue! 
For more reproducible builds (and because we didn't want to rely on a tutorial-repo) we copied the start.sh from the repo into our own repo.
This was an older version, namely from commit e049455756d69d5f4f0664109b4be78772b64179.
Now we ran again with the current version from commit 25d332c8872462a0ea75e039555096955c94c7cd and everything works.

I will close this ticket and would suggest @qriticual to check your start.sh and if that doesn't help, open another ticket :) 

Cheers for helping out and pointing us to the right direction!
